README.md
Automatic Persisted Queries
---

## Problem to solve
Unlike REST APIs that use a fixed URL to load data, GraphQL provides a rich query language that can be used to express the shape of application data requirements.  This is a marvellous advancement in technology, but it comes at a cost: GraphQL query strings are often much longer than REST URLS — in some cases by many kilobytes.

In practice we've seen GraphQL query sizes ranging well above 10 KB *just for the query text*.  This is actually significant overhead when compared with a simple URL of 50-100 characters.  When paired with the fact that the uplink speed from the client is typically the most bandwidth-constrained part of the chain, large queries can become bottlenecks for client performance.

Automatic Persisted Queries solves this problem by sending a generated ID instead of the query text as the request.

For more information about this solution, read [this article announcing Automatic Persisted Queries](https://blog.apollographql.com/improve-graphql-performance-with-automatic-persisted-queries-c31d27b8e6ea).

## How it works
1. When the client makes a query, it will optimistically send a short (64-byte) cryptographic hash instead of the full query text.
2. If the backend recognizes the hash, it will retrieve the full text of the query and execute it.
3. If the backend doesn't recogize the hash, it will ask the client to send the hash and the query text to it can store them mapped together for future lookups. During this request, the backend will also fulfill the data request.

This library is a client implementation for use with Apollo Client by using custom Apollo Link.

## Installation

`npm install @apollo/link-persisted-queries --save`

## Usage

The persisted query link requires using the `http-link`. The easiest way to use them together to to concat them into a single link.

```js
import { createPersistedQueryLink } from "@apollo/link-persisted-queries";
import { createHttpLink, InMemoryCache, ApolloClient } from "@apollo/client";

// use this with Apollo Client
const link = createPersistedQueryLink().concat(createHttpLink({ uri: "/graphql" }));
const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: link,
});
```

Thats it! Now your client will start sending query signatures instead of the full text resulting in improved network performance!

#### Options
The createPersistedQueryLink function takes an optional object with configuration. Currently the only supported configutations are a key called `generateHash` which receives the query and returns the hash, a function to conditionally disabled sending persisted queries on error
- `generateHash`: a function that takes the query document and returns the hash. If not provided, `generateHash` defaults to a fast implementation of sha256 + hex digest.
- `useGETForHashedQueries`: set to `true` to use the HTTP `GET` method when sending the hashed version of queries (but not for mutations). `GET` requests require `apollo-link-http` 1.4.0 or newer, and are not compatible with `apollo-link-batch-http`.
> If you want to use `GET` for non-mutation queries whether or not they are hashed, pass `useGETForQueries: true` option to `createHttpLink` from `apollo-link-http` instead. If you want to use `GET` for all requests, pass `fetchOptions: {method: 'GET'}` to `createHttpLink`.
- `disable`: a function which takes an ErrorResponse (see below) and returns a boolean to disable any future persited queries for that session. This defaults to disabling on `PersistedQueryNotSupported` or a 400 or 500 http error

**ErrorResponse**
The argument that the optional `disable` function is given is an object with the following keys:
- `operation`: The Operation that errored (contains query, variables, operationName, and context)
- `response`: The Execution of the reponse (contains data and errors as well extensions if sent from the server)
- `graphQLErrors`: An array of errors from the GraphQL endpoint
- `networkError`: any error during the link execution or server response

*Note*: `networkError` is the value from the downlink's `error` callback. In most cases, `graphQLErrors` is the `errors` field of the result from the last `next` call. A `networkError` can contain additional fields, such as a GraphQL object in the case of a failing HTTP status code from `apollo-link-http`. In this situation, `graphQLErrors` is an alias for `networkError.result.errors` if the property exists.

## Apollo Engine
Apollo Engine supports receiving and fulfulling Automatic Persisted Queries. Simply adding this link into your client app will improve your network response times when using Apollo Engine.


### Protocol
Automatic Persisted Queries are made up of three parts: the query signature, error responses, and the negotiaion protocol.

**Query Signature**
The query signature for Automatic Persisted Queries is sent along the extensions field of a request from the client. This is a transport independent way to send extra information along with the operation.

```js
{
  operationName: 'MyQuery',
  variables: null,
  extensions: {
    persistedQuery: {
      version: 1,
      sha256Hash: hashOfQuery
    }
  }
}
```

When sending an Automatic Persisted Query, the client ommits the `query` field normally present, and instead sends an extension field with a `persistedQuery` object as shown above. The hash algorithm defaults to a `sha256` hash of the query string.

If the client needs to register the hash, the query signature will be the same but include the full query text like so:

```js
{
  operationName: 'MyQuery',
  variables: null,
  query: `query MyQuery { id }`,
  extensions: {
    persistedQuery: {
      version: 1,
      sha256Hash: hashOfQuery
    }
  }
}
```

This should only happen once across all clients when a new query is introduced into your application.

**Error Responses**
When the initial query signature is received by a backend, if it is unable to find the hash previously stored, it must send back the following response signature:

```js
{
  errors: [
    { message: 'PersistedQueryNotFound' }
  ]
}
```

If the backend doesn't support Automatic Persisted Queries, or does not want to support it for that particular client, it can send back the following which will tell the client to stop trying to send hashes all together:

```
{
  errors: [
    { message: 'PersistedQueryNotSupported' }
  ]
}
```

**Negotiation Protocol**
In order to support Automatic Persisted Queries, the client and server must follow the negotiaion steps as outlined here:

*Happy Path*
1. Client sends query signature with no `query` field
2. Server looks up query based on hash, if found, it resolves the data
3. Client receives data and completes request

*Missing hash path*
1. Client sends query signature with no `query` field
2. Server looks up query based on hash, none is found
3. Server responds with NotFound error response
4. Client sends both hash and query string to Server
5. Server fulfills response and saves query string + hash for future lookup
6. Client receives data and completes request

### Build time generation
If you want to avoid hashing in the browser, you can use a build script to include the hash as part of the request. Then you pass a function to retrieve that hash when the operation is run. This works well with projects like [this](https://github.com/leoasis/graphql-persisted-document-loader) which uses webpack to generate the hashes at build time.

If you use the above loader, you can pass `{ generateHash: ({ documentId }) => documentId }` to the `createPersistedQueryLink` call.


CHANGELOG.md
# Change log

### 1.0.0

> **Note:** As of 1.0.0, this package is now published as `@apollo/link-persisted-queries`.

- Updated to use `@apollo/client` 3.  <br/>
  [@hwillson](https://github.com/hwillson) in [#48](https://github.com/apollographql/apollo-link-persisted-queries/pull/48)

### 0.2.0
- New option: useGETForHashedQueries
- Improve error checking
- Never retry a single request more than once, even with network errors

### 0.1.0
- handle network errors correctly for 400 and 500

### 0.0.1-beta.3
- change default error handling to only disable on 400 error, not >= 500

### 0.0.1-beta.2
- improve error handling through 500s and optional function to disable

### 0.0.1-beta.1
- support hash lookup function for build time generation

### 0.0.1-beta.0
- added memoziation to hash generation based on equality of ASTs

### 0.0.1-alpha.4
- improved failsafe if server doesn't support persisted queries

### 0.0.1-alpha.3
- change key to sha256Hash that is hex encoded

### 0.0.1-alpha.2
- change from type to message for error identification

### 0.0.1-alpha.1
- if server doesn't support PQ, don't try anymore

### 0.0.1-alpha.0
- initial release


src/__tests__/index.ts
import {
  gql,
  ApolloLink,
  execute,
  Observable,
  createHttpLink,
} from '@apollo/client';
import { sha256 } from 'js-sha256';
import { print, parse } from 'graphql';
import { cloneDeep, find, times } from 'lodash';

import { createPersistedQueryLink as createPersistedQuery, VERSION } from '../';

const makeAliasFields = (fieldName, numAliases) =>
  times(numAliases, idx => `${fieldName}${idx}: ${fieldName}`).reduce(
    (aliasBody, currentAlias) => `${aliasBody}\n    ${currentAlias}`,
  );

const sleep = ms => new Promise(s => setTimeout(s, ms));
export const query = gql`
  query Test($id: ID!) {
    foo(id: $id) {
      bar
      ${makeAliasFields('title', 1000)}
    }
  }
`;

export const shortQuery = gql`
  query Test($id: ID!) {
    foo(id: $id) {
      bar
    }
  }
`;

export const variables = { id: 1 };
export const queryString = print(query);
export const hash = sha256
  .create()
  .update(queryString)
  .hex();

export const shortQueryString = print(shortQuery);
export const shortHash = sha256
  .create()
  .update(shortQueryString)
  .hex();

// support buildtime hash generation
query.documentId = hash;
export const data = {
  foo: { bar: true },
};
export const response = JSON.stringify({ data });
const errors = [{ message: 'PersistedQueryNotFound' }];
const giveUpErrors = [{ message: 'PersistedQueryNotSupported' }];
const multipleErrors = [...errors, { message: 'not logged in' }];
const errorResponse = JSON.stringify({ errors });
const giveUpResponse = JSON.stringify({ errors: giveUpErrors });
const multiResponse = JSON.stringify({ errors: multipleErrors });

const mockObserver = jest.fn(observer => {
  setTimeout(() => {
    observer.next({ data });
    observer.complete();
  }, 10);
});
const mockApolloLink = new ApolloLink(() => new Observable(mockObserver));

const mockErrorObserver = jest.fn(observer => {
  setTimeout(() => {
    observer.error(new Error('something went wrong!'));
  }, 10);
});
const mockErroredLink = new ApolloLink(() => new Observable(mockErrorObserver));

describe('happy path', () => {
  beforeEach(fetch.mockReset);
  it('sends a sha256 hash of the query under extensions', done => {
    fetch.mockResponseOnce(response);
    const link = createPersistedQuery().concat(createHttpLink());

    execute(link, { query, variables }).subscribe(result => {
      expect(result.data).toEqual(data);
      const [uri, request] = fetch.mock.calls[0];
      expect(uri).toEqual('/graphql');
      expect(request.body).toBe(
        JSON.stringify({
          operationName: 'Test',
          variables,
          extensions: {
            persistedQuery: {
              version: VERSION,
              sha256Hash: hash,
            },
          },
        }),
      );
      done();
    }, done.fail);
  });
  it('sends a version along with the request', done => {
    fetch.mockResponseOnce(response);
    const link = createPersistedQuery().concat(createHttpLink());

    execute(link, { query, variables }).subscribe(result => {
      expect(result.data).toEqual(data);
      const [uri, request] = fetch.mock.calls[0];
      expect(uri).toEqual('/graphql');
      const parsed = JSON.parse(request.body);
      expect(parsed.extensions.persistedQuery.version).toBe(VERSION);
      done();
    }, done.fail);
  });
  it('memoizes between requests', done => {
    fetch.mockResponseOnce(response);
    fetch.mockResponseOnce(response);
    const link = createPersistedQuery().concat(createHttpLink());

    let start = new Date();
    execute(link, { query, variables }).subscribe(result => {
      const firstRun = new Date() - start;
      expect(result.data).toEqual(data);
      // this one should go faster becuase of memoization
      let secondStart = new Date();
      execute(link, { query, variables }).subscribe(result2 => {
        const secondRun = new Date() - secondStart;
        expect(firstRun).toBeGreaterThan(secondRun);
        expect(result2.data).toEqual(data);
        done();
      }, done.fail);
    }, done.fail);
  });
  it('supports loading the hash from other method', done => {
    fetch.mockResponseOnce(response);
    const generateHash = query => query.documentId + 'foo';
    const link = createPersistedQuery({ generateHash }).concat(
      createHttpLink(),
    );

    execute(link, { query, variables }).subscribe(result => {
      expect(result.data).toEqual(data);
      const [uri, request] = fetch.mock.calls[0];
      expect(uri).toEqual('/graphql');
      const parsed = JSON.parse(request.body);
      expect(parsed.extensions.persistedQuery.sha256Hash).toBe(
        `${query.documentId}foo`,
      );

      done();
    }, done.fail);
  });

  it('errors if unable to convert to sha256', done => {
    fetch.mockResponseOnce(response);
    const link = createPersistedQuery().concat(createHttpLink());

    execute(link, { query: '1234', variables }).subscribe(done.fail, error => {
      expect(error.message).toMatch(/Invalid AST Node/);
      done();
    });
  });
  it('unsubscribes correctly', done => {
    const delay = new ApolloLink(() => {
      return new Observable(ob => {
        setTimeout(() => {
          ob.next({ data });
          ob.complete();
        }, 100);
      });
    });
    const link = createPersistedQuery().concat(delay);

    const sub = execute(link, { query, variables }).subscribe(
      done.fail,
      done.fail,
      done.fail,
    );

    setTimeout(() => {
      sub.unsubscribe();
      done();
    }, 10);
  });
});
describe('failure path', () => {
  beforeEach(fetch.mockReset);
  it('correctly identifies the error shape from the server', done => {
    fetch.mockResponseOnce(errorResponse);
    fetch.mockResponseOnce(response);
    const link = createPersistedQuery().concat(createHttpLink());

    execute(link, { query, variables }).subscribe(result => {
      expect(result.data).toEqual(data);
      const [_, failure] = fetch.mock.calls[0];
      expect(JSON.parse(failure.body).query).not.toBeDefined();
      const [uri, success] = fetch.mock.calls[1];
      expect(JSON.parse(success.body).query).toBe(queryString);
      expect(
        JSON.parse(success.body).extensions.persistedQuery.sha256Hash,
      ).toBe(hash);
      done();
    }, done.fail);
  });
  it('sends GET for the first response only with useGETForHashedQueries', done => {
    fetch.mockResponseOnce(errorResponse);
    fetch.mockResponseOnce(response);
    const link = createPersistedQuery({ useGETForHashedQueries: true }).concat(
      createHttpLink(),
    );

    execute(link, { query, variables }).subscribe(result => {
      expect(result.data).toEqual(data);
      const [_, failure] = fetch.mock.calls[0];
      expect(failure.method).toBe('GET');
      expect(failure.body).not.toBeDefined();
      const [uri, success] = fetch.mock.calls[1];
      expect(success.method).toBe('POST');
      expect(JSON.parse(success.body).query).toBe(queryString);
      expect(
        JSON.parse(success.body).extensions.persistedQuery.sha256Hash,
      ).toBe(hash);
      done();
    }, done.fail);
  });
  it('does not try again after receiving NotSupported error', done => {
    fetch.mockResponseOnce(giveUpResponse);
    fetch.mockResponseOnce(response);

    // mock it again so we can verify it doesn't try anymore
    fetch.mockResponseOnce(response);
    const link = createPersistedQuery().concat(createHttpLink());

    execute(link, { query, variables }).subscribe(result => {
      expect(result.data).toEqual(data);
      const [_, failure] = fetch.mock.calls[0];
      expect(JSON.parse(failure.body).query).not.toBeDefined();
      const [uri, success] = fetch.mock.calls[1];
      expect(JSON.parse(success.body).query).toBe(queryString);
      expect(JSON.parse(success.body).extensions).toBeUndefined();
      execute(link, { query, variables }).subscribe(secondResult => {
        expect(secondResult.data).toEqual(data);

        const [uri, success] = fetch.mock.calls[2];
        expect(JSON.parse(success.body).query).toBe(queryString);
        expect(JSON.parse(success.body).extensions).toBeUndefined();
        done();
      }, done.fail);
    }, done.fail);
  });

  it('works with multiple errors', done => {
    fetch.mockResponseOnce(multiResponse);
    fetch.mockResponseOnce(response);
    const link = createPersistedQuery().concat(createHttpLink());

    execute(link, { query, variables }).subscribe(result => {
      expect(result.data).toEqual(data);
      const [_, failure] = fetch.mock.calls[0];
      expect(JSON.parse(failure.body).query).not.toBeDefined();
      const [uri, success] = fetch.mock.calls[1];
      expect(JSON.parse(success.body).query).toBe(queryString);
      expect(
        JSON.parse(success.body).extensions.persistedQuery.sha256Hash,
      ).toBe(hash);
      done();
    }, done.fail);
  });
  it('handles a 500 network error and still retries', done => {
    let failed = false;
    fetch.mockResponseOnce(response);

    // mock it again so we can verify it doesn't try anymore
    fetch.mockResponseOnce(response);

    const fetcher = (...args) => {
      if (!failed) {
        failed = true;
        return Promise.resolve({
          json: () => Promise.resolve('This will blow up'),
          text: () => Promise.resolve('THIS WILL BLOW UP'),
          status: 500,
        });
      }

      return fetch(...args);
    };
    const link = createPersistedQuery().concat(
      createHttpLink({ fetch: fetcher }),
    );

    execute(link, { query, variables }).subscribe(result => {
      expect(result.data).toEqual(data);
      const [uri, success] = fetch.mock.calls[0];
      expect(JSON.parse(success.body).query).toBe(queryString);
      expect(JSON.parse(success.body).extensions).toBeUndefined();
      execute(link, { query, variables }).subscribe(secondResult => {
        expect(secondResult.data).toEqual(data);

        const [uri, success] = fetch.mock.calls[1];
        expect(JSON.parse(success.body).query).toBe(queryString);
        expect(JSON.parse(success.body).extensions).toBeUndefined();
        done();
      }, done.fail);
    }, done.fail);
  });
  it('handles a 400 network error and still retries', done => {
    let failed = false;
    fetch.mockResponseOnce(response);

    // mock it again so we can verify it doesn't try anymore
    fetch.mockResponseOnce(response);

    const fetcher = (...args) => {
      if (!failed) {
        failed = true;
        return Promise.resolve({
          json: () => Promise.resolve('This will blow up'),
          text: () => Promise.resolve('THIS WILL BLOW UP'),
          status: 400,
        });
      }

      return fetch(...args);
    };
    const link = createPersistedQuery().concat(
      createHttpLink({ fetch: fetcher }),
    );

    execute(link, { query, variables }).subscribe(result => {
      expect(result.data).toEqual(data);
      const [uri, success] = fetch.mock.calls[0];
      expect(JSON.parse(success.body).query).toBe(queryString);
      expect(JSON.parse(success.body).extensions).toBeUndefined();
      execute(link, { query, variables }).subscribe(secondResult => {
        expect(secondResult.data).toEqual(data);

        const [uri, success] = fetch.mock.calls[1];
        expect(JSON.parse(success.body).query).toBe(queryString);
        expect(JSON.parse(success.body).extensions).toBeUndefined();
        done();
      }, done.fail);
    }, done.fail);
  });

  it('only retries a 400 network error once', done => {
    let fetchCalls = 0;
    const fetcher = (...args) => {
      fetchCalls++;
      return Promise.resolve({
        json: () => Promise.resolve('This will blow up'),
        text: () => Promise.resolve('THIS WILL BLOW UP'),
        status: 400,
      });
    };
    const link = createPersistedQuery().concat(
      createHttpLink({ fetch: fetcher }),
    );

    execute(link, { query, variables }).subscribe(
      result => done.fail,
      error => {
        expect(fetchCalls).toBe(2);
        done();
      },
    );
  });
});


src/__tests__/retry.ts
import gql from 'graphql-tag';
import { execute, ApolloLink, Observable, FetchResult } from 'apollo-link';
import waitFor from 'wait-for-observables';
import { RetryLink } from 'apollo-link-retry';
import { createHttpLink } from 'apollo-link-http';

import { createPersistedQueryLink } from '../';

export const query = gql`
  query Test($id: ID!) {
    foo(id: $id) {
      bar
    }
  }
`;

export const variables = { id: 1 };
const standardError = new Error('I never work');

export const data = {
  foo: { bar: true },
};
export const response = JSON.stringify({ data });
const errors = [{ message: 'PersistedQueryNotFound' }];
const giveUpErrors = [{ message: 'PersistedQueryNotSupported' }];
const multipleErrors = [...errors, { message: 'not logged in' }];
const errorResponse = JSON.stringify({ errors });
const giveUpResponse = JSON.stringify({ errors: giveUpErrors });
const multiResponse = JSON.stringify({ errors: multipleErrors });

const fromError = (errorValue: any) => {
  return new Observable(observer => {
    observer.error(errorValue);
  });
};

describe('RetryLink', () => {
  beforeEach(fetch.mockReset);
  it('correctly allows retry link integration', done => {
    fetch.mockResponseOnce(errorResponse);
    fetch.mockResponseOnce(response);

    let count = 0;
    const fetcher = (...args) => {
      count++;
      const [_, payload] = args;
      const body = JSON.parse(payload.body);
      // on the first fetch, we return not found
      if (count === 1) expect(body.query).toBeUndefined();
      // on the second, a critical error (retried from persisted query with query in payload)
      if (count === 2) {
        expect(body.query).toBeDefined();
        return Promise.reject(new Error('server error'));
      }
      // on the third (retried with query in the payload), we return good data
      if (count === 3) expect(body.query).toBeDefined();
      if (count > 3) done.fail('fetch called too many times');
      return fetch(...args);
    };

    const retry = new RetryLink();
    const persist = createPersistedQueryLink();
    const http = createHttpLink({ fetch: fetcher });

    const link = ApolloLink.from([persist, retry, http]);

    execute(link, { query, variables }).subscribe(result => {
      expect(result.data).toEqual(data);
      expect(count).toEqual(3);
      done();
    }, done.fail);
  });
});


src/index.ts
import {
  ApolloLink,
  Observable,
  Observer,
  Operation,
} from '@apollo/client/core';
const sha256 = require('hash.js/lib/hash/sha/256');
import { print } from 'graphql/language/printer';
import {
  DefinitionNode,
  DocumentNode,
  ExecutionResult,
  GraphQLError,
} from 'graphql';

export const VERSION = 1;

export interface ErrorResponse {
  graphQLErrors?: readonly GraphQLError[];
  networkError?: Error;
  response?: ExecutionResult;
  operation: Operation;
}

namespace PersistedQueryLink {
  export type Options = {
    generateHash?: (document: DocumentNode) => string;
    disable?: (error: ErrorResponse) => boolean;
    useGETForHashedQueries?: boolean;
  };
}

export const defaultGenerateHash = (query: DocumentNode): string =>
  sha256()
    .update(print(query))
    .digest('hex');

export const defaultOptions = {
  generateHash: defaultGenerateHash,
  disable: ({ graphQLErrors, operation }: ErrorResponse) => {
    // if the server doesn't support persisted queries, don't try anymore
    if (
      graphQLErrors &&
      graphQLErrors.some(
        ({ message }) => message === 'PersistedQueryNotSupported',
      )
    ) {
      return true;
    }

    const { response } = operation.getContext();
    // if the server responds with bad request
    // apollo-server responds with 400 for GET and 500 for POST when no query is found
    if (
      response &&
      response.status &&
      (response.status === 400 || response.status === 500)
    ) {
      return true;
    }

    return false;
  },
  useGETForHashedQueries: false,
};

function definitionIsMutation(d: DefinitionNode) {
  return d.kind === 'OperationDefinition' && d.operation === 'mutation';
}

// Note that this also returns true for subscriptions.
function operationIsQuery(operation: Operation) {
  return !operation.query.definitions.some(definitionIsMutation);
}

const { hasOwnProperty } = Object.prototype;
const hashesKeyString = '__createPersistedQueryLink_hashes';
const hashesKey =
  typeof Symbol === 'function' ? Symbol.for(hashesKeyString) : hashesKeyString;
let nextHashesChildKey = 0;

export const createPersistedQueryLink = (
  options: PersistedQueryLink.Options = {},
) => {
  const { generateHash, disable, useGETForHashedQueries } = Object.assign(
    {},
    defaultOptions,
    options,
  );
  let supportsPersistedQueries = true;

  const hashesChildKey = 'forLink' + nextHashesChildKey++;
  function getQueryHash(query: DocumentNode): string {
    if (!query || typeof query !== 'object') {
      // If the query is not an object, we won't be able to store its hash as
      // a property of query[hashesKey], so we let generateHash(query) decide
      // what to do with the bogus query.
      return generateHash(query);
    }
    if (!hasOwnProperty.call(query, hashesKey)) {
      Object.defineProperty(query, hashesKey, {
        value: Object.create(null),
        enumerable: false,
      });
    }
    const hashes = (query as any)[hashesKey];
    return hasOwnProperty.call(hashes, hashesChildKey)
      ? hashes[hashesChildKey]
      : (hashes[hashesChildKey] = generateHash(query));
  }

  return new ApolloLink((operation, forward) => {
    if (!forward) {
      throw new Error(
        'PersistedQueryLink cannot be the last link in the chain.',
      );
    }

    const { query } = operation;

    let hashError: any;
    if (supportsPersistedQueries) {
      try {
        operation.extensions.persistedQuery = {
          version: VERSION,
          sha256Hash: getQueryHash(query),
        };
      } catch (e) {
        hashError = e;
      }
    }

    return new Observable((observer: Observer<ExecutionResult>) => {
      if (hashError) {
        observer.error!(hashError);
        return;
      }

      let subscription: ZenObservable.Subscription;
      let retried = false;
      let originalFetchOptions: any;
      let setFetchOptions = false;
      const retry = (
        {
          response,
          networkError,
        }: { response?: ExecutionResult; networkError?: Error },
        cb: () => void,
      ) => {
        if (!retried && ((response && response.errors) || networkError)) {
          retried = true;

          const disablePayload = {
            response,
            networkError,
            operation,
            graphQLErrors: response ? response.errors : undefined,
          };

          // if the server doesn't support persisted queries, don't try anymore
          supportsPersistedQueries = !disable(disablePayload);

          // if its not found, we can try it again, otherwise just report the error
          if (
            (response &&
              response.errors &&
              response.errors.some(
                ({ message }: { message: string }) =>
                  message === 'PersistedQueryNotFound',
              )) ||
            !supportsPersistedQueries
          ) {
            // need to recall the link chain
            if (subscription) subscription.unsubscribe();
            // actually send the query this time
            operation.setContext({
              http: {
                includeQuery: true,
                includeExtensions: supportsPersistedQueries,
              },
            });
            if (setFetchOptions) {
              operation.setContext({ fetchOptions: originalFetchOptions });
            }
            subscription = forward(operation).subscribe(handler);

            return;
          }
        }
        cb();
      };
      const handler = {
        next: (response: ExecutionResult) => {
          retry({ response }, () => observer.next!(response));
        },
        error: (networkError: Error) => {
          retry({ networkError }, () => observer.error!(networkError));
        },
        complete: observer.complete!.bind(observer),
      };

      // don't send the query the first time
      operation.setContext({
        http: {
          includeQuery: !supportsPersistedQueries,
          includeExtensions: supportsPersistedQueries,
        },
      });

      // If requested, set method to GET if there are no mutations. Remember the
      // original fetchOptions so we can restore them if we fall back to a
      // non-hashed request.
      if (
        useGETForHashedQueries &&
        supportsPersistedQueries &&
        operationIsQuery(operation)
      ) {
        operation.setContext(
          ({ fetchOptions = {} }: { fetchOptions: Record<string, any> }) => {
            originalFetchOptions = fetchOptions;
            return {
              fetchOptions: Object.assign({}, fetchOptions, { method: 'GET' }),
            };
          },
        );
        setFetchOptions = true;
      }

      subscription = forward(operation).subscribe(handler);

      return () => {
        if (subscription) subscription.unsubscribe();
      };
    });
  });
};
